File: examples/lan-bridge/mobile/lib/protocol.dart
import 'dart:typed_data';

const protocolVersion = 1;
const inputMove = 0x01;
const inputButton = 0x02;
const inputScroll = 0x03;
const inputText = 0x04;

Uint8List encodeMove(int dx, int dy) {
  final data = ByteData(8);
  data.setUint8(0, 0x48); data.setUint8(1, 0x44); data.setUint8(2, protocolVersion); data.setUint8(3, inputMove);
  data.setInt16(4, dx.clamp(-32768, 32767)); data.setInt16(6, dy.clamp(-32768, 32767));
  return data.buffer.asUint8List();
}

Uint8List encodeButton({required bool right}) => Uint8List.fromList([0x48, 0x44, protocolVersion, inputButton, right ? 2 : 1]);
Uint8List encodeScroll(int dx, int dy) { final data = ByteData(8); data.setUint8(0, 0x48); data.setUint8(1, 0x44); data.setUint8(2, protocolVersion); data.setUint8(3, inputScroll); data.setInt16(4, dx.clamp(-32768, 32767)); data.setInt16(6, dy.clamp(-32768, 32767)); return data.buffer.asUint8List(); }
Uint8List encodeText(String text) => Uint8List.fromList([0x48, 0x44, protocolVersion, inputText, ...text.codeUnits]);

File: examples/lan-bridge/desktop/protocol.mjs
export const SERVICE_TYPE = 'halodeck';
export const PROTOCOL_VERSION = 1;
export const PORT = 47777;

export const INPUT = Object.freeze({
  move: 0x01,
  button: 0x02,
  scroll: 0x03,
  text: 0x04,
});

export function encodeMove(dx, dy) {
  const packet = new ArrayBuffer(8);
  const view = new DataView(packet);
  view.setUint8(0, 0x48); view.setUint8(1, 0x44); view.setUint8(2, PROTOCOL_VERSION);
  view.setUint8(3, INPUT.move); view.setInt16(4, clampInt16(dx)); view.setInt16(6, clampInt16(dy));
  return packet;
}

export function decodeInput(data) {
  const bytes = Buffer.from(data);
  if (bytes.length < 4 || bytes[0] !== 0x48 || bytes[1] !== 0x44 || bytes[2] !== PROTOCOL_VERSION) return null;
  const type = bytes[3];
  if (type === INPUT.move && bytes.length >= 8) return { type, dx: bytes.readInt16BE(4), dy: bytes.readInt16BE(6) };
  if (type === INPUT.button && bytes.length >= 5) return { type, button: bytes[4] };
  if (type === INPUT.scroll && bytes.length >= 8) return { type, dx: bytes.readInt16BE(4), dy: bytes.readInt16BE(6) };
  if (type === INPUT.text) return { type, text: bytes.subarray(4).toString('utf8') };
  return null;
}

function clampInt16(value) { return Math.max(-32768, Math.min(32767, Math.round(value))); }
